### DISCUSSION TITLE
Challenge 9 Help needed!!!

### DISCUSSION THREAD

**Discussion by Abelaby** [Cyfrin/foundry-full-course-cu#386] [Q&A]
Hey , I ve been trying to solve the challenge 9 for some days and i believe i have found a way but I think I would require large amounts of gas.
Im currently at Lesson 12 and have completed all other challenges up until now.
I have gone through almost all discussions in about this challenge.
So help is much appreciated.

**Comment by Jaseempk** [comment]
What help you need bro?

**Reply by Abelaby** [reply]
Any hint on solving the challenge

**Comment by asobiesk** [comment]
Can you please describe your idea so we know at which point you're stuck?

**Reply by Abelaby** [reply]
I apologize in advance if im spoiling the challenge for anyone reading this!!!






It might be dumb but I thought of having another smart contract try all the possible variations to get the correct randomGuess value and passing it to the challenge contract and stop when the transaction doesnt revert, Im thinking of limiting the gas usage by passing in some range but I still think this is not the route to take.

**Reply by asobiesk** [reply]
[SPOILERS BELOW]

You don't have to try all possible variations. The "random" `correctAnswer` value is not really random at all.

```solidity
 uint256 correctAnswer =
            uint256(keccak256(abi.encodePacked(msg.sender, block.prevrandao, block.timestamp))) % 100000;
```

This is how the LessonNine contract calculates the correct answer. You know what msg.sender will be since it will be the address your contract. The two remaining "random" variables are read from the block itself. If the LessonNine contract can read them from the block, so can your contract :)

**Reply by Abelaby** [reply]
I believe im doing it right now but the transaction is reverting.
Can i share my script with you?

**Reply by Abelaby** [reply]
Getting this error while reverting:
```
Fail with error 'ERC721: transfer to non ERC721Receiver implementer'
```

**Reply by asobiesk** [reply]
ERC721 is actually checking if the contract it's about to send the token to is aware of the ERC721 infrastructure. If it wasn't the case, tokens could be sent to contracts that do not provide any way to retrieve them afterwards, thus locking them forever. This error is related to that safety check. Look for the ERC721 receiver interface code to see what your contract is missing.

**Reply by Niferu** [reply]
Hello guys,

I have solved this challenge partially as I received NFT to my contract, but I'm unable to transfer it... I'm not sure why is that, did anyone had similar issues?

![image](https://github.com/Cyfrin/foundry-full-course-f23/assets/110537615/7f7eb923-aa73-47e6-889f-a1afcfdaa5c3)

**Reply by asobiesk** [reply]
Hi,
if you have solved that with the contract, that means you surely implemented this special interface method that is needed for receiving the NFT (don't want to mention the name and spoil the challenge). The key is you need to make that method transfer the NFT to your EOA. You can alternatively implement other method that will perform the transfer to your account. If you have solved the challenge with a contract that does not implement any way to transfer the NFT to your account, it is stuck. You have to deploy another one and solve challenge again :(

**Reply by Niferu** [reply]
Hi @asobiesk

Can I quick talk to you via discord for example?

I dont want to spoil challenge here...

The point is I have implemented necessary interface to transfer this NFT, but it doesnt allow me to do so it does not recognize tokenId I believe

**Reply by asobiesk** [reply]
Sure :)

**Reply by Niferu** [reply]
@asobiesk 

Please add me then, so I can DM you, thanks a lot!

**Reply by Ynyesto** [reply]
Hi, guys! I'm also stuck on this. My first question would be: can't this be solved from Foundry with a script contract using my private key for the broadcasts? (vm.startBroadcast(vm.envUint("PRIVATE_KEY")), as seen in lesson 9) I have tried this way and it just makes sense to me, the answer gets calculated in the script contract and sent to the solveChallenge function. I do get LessonNine__WrongValue(), but I don't get why... Does the block change in the time it takes to call solveChallenge since the guess is calculated??? I'm at my wits' end on this :-/ I'd very much appreciate any hints!

**Comment by basefee** [comment]
Spent 4-5 hours on this challenge, by far it was the most difficult one for me.
I did it after completing Lesson 11. So, used some function selector but definitely can be done without it.
If @PatrickAlphaC agrees, i would share my contract and spoil this challenge. Be sure to spend ample time on it.

Comment if you want some hints.

**Reply by Abelaby** [reply]
Thank you mate. However I completed the challenge, thanks @asobiesk for you tip on ERC721 receiver interface. 
I agree that this was the most hardest challenge so far and I actually spend a lot of time trying to solve this using cast and I was unsuccessful in finding a  solution without deploying a contract.


With that said, I'm open  to suggestions and spoils on how to solve this challenge using cast.

**Reply by PatrickAlphaC** [reply]
No spoilers here please :) 

Feel free to write an article about how you solved each challenge, though, and share that with your friends.

**Comment by tbrannt** [comment]
There's a way to solve this challenge without deploying a smart contract. I just did that. If anyone has some question I can give some hints.

**Reply by scurenza** [reply]
I'm trying to resolve this challenge without deploying smart contract but till now it was a waste of SepoliaETH. Can i DM you on Discord?

**Reply by tbrannt** [reply]
hey, I'm not on discord. In this thread are some hints: https://github.com/Cyfrin/foundry-full-course-cu/discussions/385#discussioncomment-7826003

**Reply by scurenza** [reply]
Thanks a lot!

**Comment by sagoto80** [comment]
can anyone help on this #1181

**Comment by mirmohmmadluqman** [comment]
Hey @scurenza, man, I feel you—burning Sepolia ETH on Challenge 9 is rough, especially when you’re so close! I’ve been grinding the Cyfrin Foundry course (Patrick Collins’ epic run) and sifting through #386 and related threads. You’re on the no-contract-deploy path @tbrannt mentioned—smart call, but it’s a tough nut to crack without spoilers.

This likely involves ERC721 NFT mechanics—maybe a reward transfer you need to nail perfectly. That gas drain? Probably from misfired calls—let’s zero in smartly. Since @PatrickAlphaC keeps it spoiler-free, here’s the best short boost:

- **Cast Smarts**: If it’s a dynamic guess (block value/hash), use `cast --block-number` or fork with `anvil --fork-url $SEPOLIA_RPC_URL`, test locally, then script with `vm.startBroadcast` in Foundry. Add a 1-2 block buffer to dodge tx delays.
- **ERC721 Fix**: Ensure your receiver callback (that safeTransferFrom hook) forwards the NFT to your EOA with `transferFrom`. Test with a mock ERC721 locally.
- **Low-Level Hack**: Use `cast calldata` or keccak for selectors (Lesson 11 style) and send via `cast send --data 0x...` to avoid gas waste.
- **Debug Fast**: Fork with `forge test --fork-url $RPC`, trace with `forge debug`, and compare expected vs. actual values.

If ETH keeps vanishing, deploy a minimal OpenZeppelin ERC721Receiver contract—practical, not cheating, and many did it. Check @tbrannt’s #385 hints. Share error logs (no full code), and we’ll nudge you to that NFT win! You’ve got this! 🚀
